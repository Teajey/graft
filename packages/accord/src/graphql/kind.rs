pub use macros::Kind;
use serde::{de, Deserialize};

use super::query::Name;

pub trait Kind: for<'de> Deserialize<'de> {}

pub fn visit_map<'de, A>(invalid_kind_msg: &'static str, mut map: A) -> Result<String, A::Error>
where
    A: serde::de::MapAccess<'de>,
{
    let mut kind = None;
    let mut value = None;

    while let Some(key) = map.next_key::<String>()? {
        match key.as_str() {
            "kind" => {
                if kind.is_some() {
                    return Err(de::Error::duplicate_field("kind"));
                }
                kind = Some(map.next_value::<String>()?);
            }
            "value" => {
                if value.is_some() {
                    return Err(de::Error::duplicate_field("value"));
                }
                value = Some(map.next_value()?);
            }
            x => return Err(de::Error::unknown_field(x, &["kind", "value"])),
        }
    }

    let Some("Name") = kind.as_deref() else {
        return Err(de::Error::custom(invalid_kind_msg));
    };

    let Some(value) = value else {
        return Err(de::Error::missing_field("value"))
    };

    Ok(value)
}

impl crate::graphql::kind::Kind for super::query::Name {}
struct NameVisitor;
impl<'de> serde::de::Visitor<'de> for NameVisitor {
    type Value = Name;

    fn expecting(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(
            f,
            "a map {{ kind: \"{}\", value: <some name> }}",
            stringify!(Name)
        )
    }

    fn visit_map<A>(self, map: A) -> Result<Self::Value, A::Error>
    where
        A: serde::de::MapAccess<'de>,
    {
        let value = crate::graphql::kind::visit_map("\"kind\" must be string \"#name\"", map)?;
        Ok(Name(value))
    }
}
impl<'de> Deserialize<'de> for Name {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        deserializer.deserialize_map(NameVisitor)
    }
}
