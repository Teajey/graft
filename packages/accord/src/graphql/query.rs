mod from_ast;

use serde::Deserialize;

use super::kind::Kind;

#[derive(PartialEq, Debug, Kind)]
pub struct Name(String);

#[derive(Deserialize, Debug)]
pub struct Directive {
    name: Name,
    arguments: Vec<Argument>,
}

#[derive(Deserialize, Debug)]
pub struct ObjectField {
    name: Name,
    value: Value,
}

#[derive(Deserialize, Debug)]
pub struct Variable {
    name: Name,
}

#[derive(Deserialize, Debug)]
#[serde(tag = "kind")]
pub enum Value {
    Variable(Variable),
    #[serde(rename = "IntValue")]
    Int {
        value: String,
    },
    #[serde(rename = "FloatValue")]
    Float {
        value: f64,
    },
    #[serde(rename = "ObjectValue")]
    Object {
        fields: Vec<ObjectField>,
    },
    #[serde(rename = "StringValue")]
    String {
        value: String,
        block: bool,
    },
    #[serde(rename = "ListValue")]
    List {
        values: Vec<Value>,
    },
    #[serde(rename = "BooleanValue")]
    Boolean {
        value: bool,
    },
    #[serde(rename = "NullValue")]
    Null,
    #[serde(rename = "EnumValue")]
    Enum {
        value: String,
    },
}

#[derive(Deserialize, Debug)]
pub struct Argument {
    name: Name,
    value: Value,
}

#[derive(Deserialize, Debug)]
pub struct NamedType {
    name: Name,
}

#[derive(Deserialize, Debug)]
#[serde(tag = "kind")]
pub enum Type {
    #[serde(rename = "NamedType")]
    Named(NamedType),
    #[serde(rename = "NonNullType")]
    NonNull { value: Box<Type> },
    #[serde(rename = "ListType")]
    List { value: Box<Type> },
}

#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct VariableDefinition {
    variable: Variable,
    #[serde(rename = "type")]
    of_type: Type,
    default_value: Option<Value>,
    directives: Vec<Directive>,
}

#[derive(Deserialize, Debug)]
#[serde(tag = "kind")]
pub enum Selection {
    #[serde(rename_all = "camelCase")]
    Field {
        alias: Option<Name>,
        name: Name,
        arguments: Vec<Argument>,
        directives: Vec<Directive>,
        selection_set: Option<SelectionSet>,
    },
    #[serde(rename_all = "camelCase")]
    InlineFragment {
        type_condition: Option<NamedType>,
        directives: Vec<Directive>,
        selection_set: SelectionSet,
    },
    #[serde(rename_all = "camelCase")]
    FragmentSpread {
        name: Name,
        directives: Vec<Directive>,
    },
}

#[derive(Deserialize, Debug)]
pub struct SelectionSet {
    selections: Vec<Selection>,
}

#[derive(Deserialize, Debug)]
#[serde(tag = "operation")]
#[serde(rename_all = "camelCase")]
pub enum Operation {
    SelectionSet(SelectionSet),
    #[serde(rename_all = "camelCase")]
    Query {
        name: Option<Name>,
        variable_definitions: Vec<VariableDefinition>,
        directives: Vec<Directive>,
        selection_set: SelectionSet,
    },
    #[serde(rename_all = "camelCase")]
    Mutation {
        name: Option<Name>,
        variable_definitions: Vec<VariableDefinition>,
        directives: Vec<Directive>,
        selection_set: SelectionSet,
    },
    #[serde(rename_all = "camelCase")]
    Subscription {
        name: Option<Name>,
        variable_definitions: Vec<VariableDefinition>,
        directives: Vec<Directive>,
        selection_set: SelectionSet,
    },
}

#[derive(Deserialize, Debug)]
#[serde(tag = "kind")]
pub enum Definition {
    #[serde(rename = "OperationDefinition")]
    Operation(Operation),
    #[serde(rename = "FragmentDefinition", rename_all = "camelCase")]
    Fragment {
        name: Name,
        type_condition: NamedType,
        directives: Vec<Directive>,
        selection_set: SelectionSet,
    },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deserialize() {
        let json_str = include_str!("../../fixtures/kitchen-sink_query.json");

        let doc: serde_json::Value = serde_json::from_str(json_str).expect("must be valid json");

        let doc: Vec<Definition> = serde_json::from_value(doc["definitions"].clone())
            .expect("kitchen sink query must be deserializable");

        insta::assert_debug_snapshot!(doc);
    }

    #[test]
    fn deserialize_name() {
        let json = serde_json::json!({
            "kind": "Name",
            "value": "queryName",
        });

        let result: Result<Name, _> = serde_json::from_value(json);

        match result {
            Err(err) => panic!("Deserialize failure {err}"),
            Ok(name) => assert_eq!(Name("queryName".to_owned()), name),
        };
    }

    #[test]
    fn deserialize_variable_definition() {
        let json = serde_json::json!({
          "kind": "VariableDefinition",
          "variable": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "input"
            }
          },
          "type": {
            "kind": "NamedType",
            "name": {
              "kind": "Name",
              "value": "StoryLikeSubscribeInput"
            }
          },
          "directives": []
        });

        let result: Result<VariableDefinition, _> = serde_json::from_value(json);

        if let Err(err) = result {
            panic!("Deserialize failure {err}");
        }
    }
}
