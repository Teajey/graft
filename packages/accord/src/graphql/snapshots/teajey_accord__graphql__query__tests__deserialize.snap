---
source: packages/accord/src/graphql/query.rs
expression: defs
---
[
    Operation(
        Query {
            name: Some(
                Name(
                    "queryName",
                ),
            ),
            variable_definitions: [
                VariableDefinition {
                    kind: Some(
                        T,
                    ),
                    variable: Variable {
                        kind: T,
                        name: Name(
                            "foo",
                        ),
                    },
                    of_type: Named {
                        name: Name(
                            "ComplexType",
                        ),
                    },
                    default_value: None,
                    directives: [],
                },
                VariableDefinition {
                    kind: Some(
                        T,
                    ),
                    variable: Variable {
                        kind: T,
                        name: Name(
                            "site",
                        ),
                    },
                    of_type: Named {
                        name: Name(
                            "Site",
                        ),
                    },
                    default_value: Some(
                        Enum {
                            value: "MOBILE",
                        },
                    ),
                    directives: [],
                },
            ],
            directives: [],
            selection_set: SelectionSet {
                kind: Some(
                    T,
                ),
                selections: [
                    Field {
                        alias: Some(
                            Name(
                                "whoever123is",
                            ),
                        ),
                        name: Name(
                            "node",
                        ),
                        arguments: [
                            Argument {
                                kind: Some(
                                    T,
                                ),
                                name: Name(
                                    "id",
                                ),
                                value: List {
                                    values: [
                                        Int {
                                            value: "123",
                                        },
                                        Int {
                                            value: "456",
                                        },
                                    ],
                                },
                            },
                        ],
                        directives: [],
                        selection_set: Some(
                            SelectionSet {
                                kind: Some(
                                    T,
                                ),
                                selections: [
                                    Field {
                                        alias: None,
                                        name: Name(
                                            "id",
                                        ),
                                        arguments: [],
                                        directives: [],
                                        selection_set: None,
                                    },
                                    InlineFragment {
                                        type_condition: Some(
                                            NamedType {
                                                kind: Some(
                                                    T,
                                                ),
                                                name: Name(
                                                    "User",
                                                ),
                                            },
                                        ),
                                        directives: [
                                            Directive {
                                                kind: Some(
                                                    T,
                                                ),
                                                name: Name(
                                                    "defer",
                                                ),
                                                arguments: [],
                                            },
                                        ],
                                        selection_set: SelectionSet {
                                            kind: Some(
                                                T,
                                            ),
                                            selections: [
                                                Field {
                                                    alias: None,
                                                    name: Name(
                                                        "field2",
                                                    ),
                                                    arguments: [],
                                                    directives: [],
                                                    selection_set: Some(
                                                        SelectionSet {
                                                            kind: Some(
                                                                T,
                                                            ),
                                                            selections: [
                                                                Field {
                                                                    alias: None,
                                                                    name: Name(
                                                                        "id",
                                                                    ),
                                                                    arguments: [],
                                                                    directives: [],
                                                                    selection_set: None,
                                                                },
                                                                Field {
                                                                    alias: Some(
                                                                        Name(
                                                                            "alias",
                                                                        ),
                                                                    ),
                                                                    name: Name(
                                                                        "field1",
                                                                    ),
                                                                    arguments: [
                                                                        Argument {
                                                                            kind: Some(
                                                                                T,
                                                                            ),
                                                                            name: Name(
                                                                                "first",
                                                                            ),
                                                                            value: Int {
                                                                                value: "10",
                                                                            },
                                                                        },
                                                                        Argument {
                                                                            kind: Some(
                                                                                T,
                                                                            ),
                                                                            name: Name(
                                                                                "after",
                                                                            ),
                                                                            value: Variable {
                                                                                name: Name(
                                                                                    "foo",
                                                                                ),
                                                                            },
                                                                        },
                                                                    ],
                                                                    directives: [
                                                                        Directive {
                                                                            kind: Some(
                                                                                T,
                                                                            ),
                                                                            name: Name(
                                                                                "include",
                                                                            ),
                                                                            arguments: [
                                                                                Argument {
                                                                                    kind: Some(
                                                                                        T,
                                                                                    ),
                                                                                    name: Name(
                                                                                        "if",
                                                                                    ),
                                                                                    value: Variable {
                                                                                        name: Name(
                                                                                            "foo",
                                                                                        ),
                                                                                    },
                                                                                },
                                                                            ],
                                                                        },
                                                                    ],
                                                                    selection_set: Some(
                                                                        SelectionSet {
                                                                            kind: Some(
                                                                                T,
                                                                            ),
                                                                            selections: [
                                                                                Field {
                                                                                    alias: None,
                                                                                    name: Name(
                                                                                        "id",
                                                                                    ),
                                                                                    arguments: [],
                                                                                    directives: [],
                                                                                    selection_set: None,
                                                                                },
                                                                                FragmentSpread {
                                                                                    name: Name(
                                                                                        "frag",
                                                                                    ),
                                                                                    directives: [],
                                                                                },
                                                                            ],
                                                                        },
                                                                    ),
                                                                },
                                                            ],
                                                        },
                                                    ),
                                                },
                                            ],
                                        },
                                    },
                                    InlineFragment {
                                        type_condition: None,
                                        directives: [
                                            Directive {
                                                kind: Some(
                                                    T,
                                                ),
                                                name: Name(
                                                    "skip",
                                                ),
                                                arguments: [
                                                    Argument {
                                                        kind: Some(
                                                            T,
                                                        ),
                                                        name: Name(
                                                            "unless",
                                                        ),
                                                        value: Variable {
                                                            name: Name(
                                                                "foo",
                                                            ),
                                                        },
                                                    },
                                                ],
                                            },
                                        ],
                                        selection_set: SelectionSet {
                                            kind: Some(
                                                T,
                                            ),
                                            selections: [
                                                Field {
                                                    alias: None,
                                                    name: Name(
                                                        "id",
                                                    ),
                                                    arguments: [],
                                                    directives: [],
                                                    selection_set: None,
                                                },
                                            ],
                                        },
                                    },
                                    InlineFragment {
                                        type_condition: None,
                                        directives: [],
                                        selection_set: SelectionSet {
                                            kind: Some(
                                                T,
                                            ),
                                            selections: [
                                                Field {
                                                    alias: None,
                                                    name: Name(
                                                        "id",
                                                    ),
                                                    arguments: [],
                                                    directives: [],
                                                    selection_set: None,
                                                },
                                            ],
                                        },
                                    },
                                ],
                            },
                        ),
                    },
                ],
            },
        },
    ),
    Operation(
        Mutation {
            name: Some(
                Name(
                    "likeStory",
                ),
            ),
            variable_definitions: [],
            directives: [],
            selection_set: SelectionSet {
                kind: Some(
                    T,
                ),
                selections: [
                    Field {
                        alias: None,
                        name: Name(
                            "like",
                        ),
                        arguments: [
                            Argument {
                                kind: Some(
                                    T,
                                ),
                                name: Name(
                                    "story",
                                ),
                                value: Int {
                                    value: "123",
                                },
                            },
                        ],
                        directives: [
                            Directive {
                                kind: Some(
                                    T,
                                ),
                                name: Name(
                                    "defer",
                                ),
                                arguments: [],
                            },
                        ],
                        selection_set: Some(
                            SelectionSet {
                                kind: Some(
                                    T,
                                ),
                                selections: [
                                    Field {
                                        alias: None,
                                        name: Name(
                                            "story",
                                        ),
                                        arguments: [],
                                        directives: [],
                                        selection_set: Some(
                                            SelectionSet {
                                                kind: Some(
                                                    T,
                                                ),
                                                selections: [
                                                    Field {
                                                        alias: None,
                                                        name: Name(
                                                            "id",
                                                        ),
                                                        arguments: [],
                                                        directives: [],
                                                        selection_set: None,
                                                    },
                                                ],
                                            },
                                        ),
                                    },
                                ],
                            },
                        ),
                    },
                ],
            },
        },
    ),
    Operation(
        Subscription {
            name: Some(
                Name(
                    "StoryLikeSubscription",
                ),
            ),
            variable_definitions: [
                VariableDefinition {
                    kind: Some(
                        T,
                    ),
                    variable: Variable {
                        kind: T,
                        name: Name(
                            "input",
                        ),
                    },
                    of_type: Named {
                        name: Name(
                            "StoryLikeSubscribeInput",
                        ),
                    },
                    default_value: None,
                    directives: [],
                },
            ],
            directives: [],
            selection_set: SelectionSet {
                kind: Some(
                    T,
                ),
                selections: [
                    Field {
                        alias: None,
                        name: Name(
                            "storyLikeSubscribe",
                        ),
                        arguments: [
                            Argument {
                                kind: Some(
                                    T,
                                ),
                                name: Name(
                                    "input",
                                ),
                                value: Variable {
                                    name: Name(
                                        "input",
                                    ),
                                },
                            },
                        ],
                        directives: [],
                        selection_set: Some(
                            SelectionSet {
                                kind: Some(
                                    T,
                                ),
                                selections: [
                                    Field {
                                        alias: None,
                                        name: Name(
                                            "story",
                                        ),
                                        arguments: [],
                                        directives: [],
                                        selection_set: Some(
                                            SelectionSet {
                                                kind: Some(
                                                    T,
                                                ),
                                                selections: [
                                                    Field {
                                                        alias: None,
                                                        name: Name(
                                                            "likers",
                                                        ),
                                                        arguments: [],
                                                        directives: [],
                                                        selection_set: Some(
                                                            SelectionSet {
                                                                kind: Some(
                                                                    T,
                                                                ),
                                                                selections: [
                                                                    Field {
                                                                        alias: None,
                                                                        name: Name(
                                                                            "count",
                                                                        ),
                                                                        arguments: [],
                                                                        directives: [],
                                                                        selection_set: None,
                                                                    },
                                                                ],
                                                            },
                                                        ),
                                                    },
                                                    Field {
                                                        alias: None,
                                                        name: Name(
                                                            "likeSentence",
                                                        ),
                                                        arguments: [],
                                                        directives: [],
                                                        selection_set: Some(
                                                            SelectionSet {
                                                                kind: Some(
                                                                    T,
                                                                ),
                                                                selections: [
                                                                    Field {
                                                                        alias: None,
                                                                        name: Name(
                                                                            "text",
                                                                        ),
                                                                        arguments: [],
                                                                        directives: [],
                                                                        selection_set: None,
                                                                    },
                                                                ],
                                                            },
                                                        ),
                                                    },
                                                ],
                                            },
                                        ),
                                    },
                                ],
                            },
                        ),
                    },
                ],
            },
        },
    ),
    Fragment {
        name: Name(
            "frag",
        ),
        type_condition: NamedType {
            kind: Some(
                T,
            ),
            name: Name(
                "Friend",
            ),
        },
        directives: [],
        selection_set: SelectionSet {
            kind: Some(
                T,
            ),
            selections: [
                Field {
                    alias: None,
                    name: Name(
                        "foo",
                    ),
                    arguments: [
                        Argument {
                            kind: Some(
                                T,
                            ),
                            name: Name(
                                "size",
                            ),
                            value: Variable {
                                name: Name(
                                    "size",
                                ),
                            },
                        },
                        Argument {
                            kind: Some(
                                T,
                            ),
                            name: Name(
                                "bar",
                            ),
                            value: Variable {
                                name: Name(
                                    "b",
                                ),
                            },
                        },
                        Argument {
                            kind: Some(
                                T,
                            ),
                            name: Name(
                                "obj",
                            ),
                            value: Object {
                                fields: [
                                    ObjectField {
                                        kind: Some(
                                            T,
                                        ),
                                        name: Name(
                                            "key",
                                        ),
                                        value: String {
                                            value: "value",
                                            block: false,
                                        },
                                    },
                                    ObjectField {
                                        kind: Some(
                                            T,
                                        ),
                                        name: Name(
                                            "block",
                                        ),
                                        value: String {
                                            value: "              block string uses \\\"\\\"\\\"          ",
                                            block: true,
                                        },
                                    },
                                ],
                            },
                        },
                    ],
                    directives: [],
                    selection_set: None,
                },
            ],
        },
    },
    Operation(
        Query {
            name: None,
            variable_definitions: [],
            directives: [],
            selection_set: SelectionSet {
                kind: Some(
                    T,
                ),
                selections: [
                    Field {
                        alias: None,
                        name: Name(
                            "unnamed",
                        ),
                        arguments: [
                            Argument {
                                kind: Some(
                                    T,
                                ),
                                name: Name(
                                    "truthy",
                                ),
                                value: Boolean {
                                    value: true,
                                },
                            },
                            Argument {
                                kind: Some(
                                    T,
                                ),
                                name: Name(
                                    "falsey",
                                ),
                                value: Boolean {
                                    value: false,
                                },
                            },
                            Argument {
                                kind: Some(
                                    T,
                                ),
                                name: Name(
                                    "nullish",
                                ),
                                value: Null,
                            },
                        ],
                        directives: [],
                        selection_set: None,
                    },
                    Field {
                        alias: None,
                        name: Name(
                            "query",
                        ),
                        arguments: [],
                        directives: [],
                        selection_set: None,
                    },
                ],
            },
        },
    ),
]
